src/mpid/ch4/shm/xpmem/xpmem_init.c

```diff
@@ -26,8 +26,14 @@ int MPIDI_XPMEM_mpi_init_hook(int rank, int size, int *n_vcis_provided, int *tag
     /* Try to share entire address space */
     MPIDI_XPMEM_global.segid = xpmem_make(0, XPMEM_MAXADDR_SIZE, XPMEM_PERMIT_MODE,
                                           MPIDI_XPMEM_PERMIT_VALUE);
-    /* 64-bit segment ID or failure(-1) */
-    MPIR_ERR_CHKANDJUMP(MPIDI_XPMEM_global.segid == -1, mpi_errno, MPI_ERR_OTHER, "**xpmem_make");
+    /* While this version of MPICH does require libxpmem to link, we don't necessarily require the
+     * kernel module to be loaded at runtime. If XPMEM is not available, disable its use via the
+     * special CVAR value. */
+    if (MPIDI_XPMEM_global.segid == -1) {
+        XPMEM_TRACE("init: xpmem_make failed. Disabling XPMEM support");
+        MPIR_CVAR_CH4_XPMEM_LMT_MSG_SIZE = -1;
+        goto fn_exit;
+    }
     XPMEM_TRACE("init: make segid: 0x%lx\n", (uint64_t) MPIDI_XPMEM_global.segid);
 
```

Comment:
Maybe we want to exchange the return value of `xpmem_make` with all local processes and disable XPMEM if any of them is `-1`. `xpmem_make` may also return `-1` if the internal call to `ioctl()` fails. I am not sure if all local processes always see the same failure. Otherwise, the sender may expect XPMEM but the receiver disables it.

Besides, if any process returns a valid segid, we need to call `xpmem_remove` at `fn_fail` because `MPIDI_XPMEM_mpi_finalize_hook` will not be invoked.

Reply:
I took care of both of these. To accomplish it, I added `MPIDU_Init_shm_allsum` as a simple allreduce for shared memory to be able to check the return value of `xpmem_make`. This reuses the infrastructure from `MPIDU_Init_shm_barrier`.